### PR TITLE
Backport of Fix up case where subscription is terminated due to ACLs changing or a snapshot restore occurring into release/1.15.x

### PR DESCRIPTION
**Pull request by hc-github-team-consul-core**
## Backport

This PR is auto-generated from #17566 to be assessed for backporting due to the inclusion of the label backport/1.15.

The below text is copied from the body of the original PR.

---

### Description

Due to swallowing and only logging xDS errors here:

https://github.com/hashicorp/consul/blob/88951bfafaa969f52235dc5b443cb64ba329660c/agent/proxycfg/state.go#L375-L386

We were getting into a state where if ACLs were modified while connected to xDS, the stream wasn't properly being re-established and instead the Consul logs were filling up with error logs such as:

> Failed to handle update from watch: kind=api-gateway proxy=default/default/gateway-d6b4d69c-krz5b service_id=default/default/gateway-d6b4d69c-krz5b id=gateway-config error="error filling agent cache: subscription closed by server, client must reset state and resubscribe"

Basically we were missing some casing for marking certain errors as `TerminalError`s in proxycfg-glue, this fixes that.

### Testing & Reproduction steps

Get a proxy and/or gateway to connect up to xDS and then modify the token ACLs that are associated with it, you should see a bunch of these in the logs.

### PR Checklist

* [ ] updated test coverage
* [ ] external facing docs updated
* [ ] appropriate backport labels added
* [ ] not a security concern

---

<details>
<summary> Overview of commits </summary>

  - 82e7d4fe18d450b2d43d476b704dc3ef1d155184  - 45008e27c356321d7c94b23b165cce1b04a5dfb3  - bdee9e3b9858ca9b654de853bcb4c77d8f14843e 

</details>

### CODE REVIEW

**Review by github-team-consul-core-pr-approver — APPROVED** [review]
Auto approved Consul Bot automated PR
